### PR TITLE
pkg/policy/rule: Optimize rule String()

### PR DESCRIPTION
**Pull request by MikeLing**
Avoiding to use  fmt.Sprintf() so that Go won't over-allocate the memory.

issue reference : cilium#19571

Result:

```
$ go test -v -run '^$' -bench 'BenchmarkRuleString' -benchmem ./pkg/policy > BenchmarkRuleString_old.txt
$ go test -v -run '^$' -bench 'BenchmarkRuleString' -benchmem ./pkg/policy > BenchmarkRuleString_new.txt
$ benchcmp BenchmarkRuleString_old.txt BenchmarkRuleString_new.txt
benchmark                 old ns/op     new ns/op     delta
BenchmarkRuleString-4     1929          1514          -21.51%

benchmark                 old allocs     new allocs     delta
BenchmarkRuleString-4     13             10             -23.08%

benchmark                 old bytes     new bytes     delta
BenchmarkRuleString-4     752           656           -12.77%
```

Signed-off-by: MikeLing <sabergeass@gmail.com>

### CODE REVIEW

File: pkg/policy/rule_test.go [hunk 1]
```diff
@@ -2575,3 +2576,40 @@ func (ds *PolicyTestSuite) TestMatches(c *C) {
...
+					{
+						ToPorts: []api.PortRule{{
+							Ports: []api.PortProtocol{
+								{Port: "3000", Protocol: api.ProtoAny},
+							},
+						}},
+					},
+				},
+			},
+		}
```
Comment (tklauser) [inline pkg/policy/rule_test.go]: I'd suggest to move the rule construction out of the loop before `b.ReportAllocs`, so it doesn't influence benchmarks (namely allocations).

Comment (MikeLing) [reply pkg/policy/rule_test.go]: Already update the code and benchmark result.

**Review by tklauser — APPROVED** [review]

**Review by christarazi — CHANGES_REQUESTED** [review]
Checkpatch job seems to be failing because the commit msg author emails differ from the sign-off inside the commit. Could you fix those up?

```
Error: ERROR:NO_AUTHOR_SIGN_OFF: Missing Signed-off-by: line by nominal patch author 'linweijie08 <linweijie08@meituan.com>'
```

**Comment by MikeLing** [comment]
> Checkpatch job seems to be failing because the commit msg author emails differ from the sign-off inside the commit. Could you fix those up?
> 
> ```
> Error: ERROR:NO_AUTHOR_SIGN_OFF: Missing Signed-off-by: line by nominal patch author 'linweijie08 <linweijie08@meituan.com>'
> ```

Already fix it. BTW, could I trigger the `tests` with `/test` by myself?

**Review by christarazi — APPROVED** [review]

**Review by kkourt — APPROVED** [review]

**Comment by christarazi** [comment]
Marking ready to merge as we have approvals and CI passed aside from known failures.

**Pull request merged**